synadm/cli/notice.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+# synadm
+# Copyright (C) 2022 Philip (a-0)
+#
+# synadm is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# synadm is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""server notice-related CLI commands
+"""
+
+import re
+import click
+from click_option_group import optgroup, MutuallyExclusiveOptionGroup
+from click_option_group import RequiredAnyOptionGroup
+
+from synadm import cli
+
+@cli.root.group()
+def notice():
+    """ Send server notices to local users
+    """
+
+@notice.command(name="send")
+@click.option("--from-file/--from-argument", "-f/-a", default=False,
+    help="""Interpret arguments as file paths instead of notice content and 
+    read content from those files""")
+@click.option("--paginate", type=int, default=100,
+    help="Sets how many users will be retrieved from the server at once")
+@click.argument("to", type=str, default=None)
+@click.argument("plain", type=str, default=None)
+@click.argument("formatted", type=str, default=None, required=False)
+@click.pass_obj
+def notice_send_cmd(helper, from_file, paginate, to, plain, formatted):
+    """Send server notices to local users.
+
+    TO - either a matrix ID (e.g. '@abc:example.com') or a regular expression 
+        as used in python (e.g. '^.*' to send to all users)
+    
+    PLAIN - plain text content of the notice
+    
+    FORMATTED - (HTML-) formatted content of the notice. If not set, PLAIN 
+        will be used.
+    """
+    if from_file:
+        with open(plain, "r") as plain_file:
+            plain_content = plain_file.read()
+        if formatted:
+            with open(formatted, "r") as formatted_file:
+                formatted_content = formatted_file.read()
+        else:
+            formatted_content = plain_content
+    else:
+        plain_content = plain
+        formatted_content = formatted
+    
+    def confirm_prompt(users):
```

Comment:
I'm getting this on my rather old Synapse installation. It might happen because the notice admin API is not present and returning a 404. We could try to catch non-supported Synapse versions early and throw a note to the user. 

```
(synadm) ~/git/a-0_synadm (notice) $ synadm notice send "blabla" @testuser1:peek-a-boo.at
Traceback (most recent call last):
  File "/home/jojo/.venvs/synadm/bin/synadm", line 11, in <module>
    load_entry_point('synadm', 'console_scripts', 'synadm')()
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/jojo/git/a-0_synadm/synadm/cli/notice.py", line 89, in notice_send_cmd
    if not confirm_prompt([to]):
  File "/home/jojo/git/a-0_synadm/synadm/cli/notice.py", line 80, in confirm_prompt
    + "\n---\nSend now?"
TypeError: must be str, not NoneType
(synadm) ~/git/a-0_synadm (notice) $
(synadm) ~/git/a-0_synadm (notice) $
(synadm) ~/git/a-0_synadm (notice) $ synadm version
python_version: 3.6.9
server_version: 1.49.2

(synadm) ~/git/a-0_synadm (notice) $
```

Reply:
Update: notice API exists since Synapse 0.30, so that' not the issue ;-) I think it's the default value of the formatted_message that is None and thus, if not given, this concatenation error is thrown.

Reply:
I missed one case of input parameters: Do not read content from files, but no formatted body supplied via CLI. In that case formatted_content was set to None, which of course causes issues. Will be fixed in my next commit.